pkg/ruler/ruler.go

```diff
@@ -488,11 +517,11 @@ func tokenForGroup(g *rulespb.RuleGroupDesc) uint32 {
 	return ringHasher.Sum32()
 }
 
-func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups, instanceAddr string, forBackup bool) (bool, error) {
+func instanceOwnsRuleGroup(r ruler, rr ring.ReadRing, g *rulespb.RuleGroupDesc, disabledRuleGroups validation.DisabledRuleGroups, instanceAddr string, forBackup bool) (bool, error) {
```

Comment:
Should we just use `*Ruler` as the method receiver here?

Reply:
Reasons for introducing an interface is to narrow down what is available for functions to consume. Existing functions such as `filterRuleGroups` is not a method to avoid accidentally using Ruler's ring directly. If I pass in Ruler struct/object as-is, then it will expose the ring to those functions which could lead to accidental use of it. This is the reason I introduced a interface so that it could narrow down the scope of what is available to these functions. Of course, someone in the future could just add a method on the interface that exposes the ring. Having said all this, I'm open to suggestions here

Reply:
I changed `filterRuleGroups` and other functions to methods on the Ruler and removed the interface